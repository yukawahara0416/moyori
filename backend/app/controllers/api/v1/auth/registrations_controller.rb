module Api
  module V1
    module Auth
      class RegistrationsController < DeviseTokenAuth::RegistrationsController
        def destroy
          if @resource
            @resource.destroy
            @resource.avatar.prune if @resource.avatar.attached?
            yield @resource if block_given?
            render_destroy_success
          else
            render_destroy_error
          end
        end

        def render_create_success
          render json: convert_to_registrations(@resource)
        end

        def render_update_success
          render json: convert_to_registrations(@resource)
        end

        private

          def sign_up_params
            params.permit(:name, :email, :password, :avatar)
          end

          def account_update_params
            params.permit(:name, :email, :avatar)
          end

          def convert_to_registrations(resource)
            {
              data: {
                id: resource.id,
                email: resource.email,
                provider: resource.provider,
                name: resource.name,
                avatar: set_avatar(resource.id),
                uid: resource.uid,
                allow_password_change: resource.allow_password_change,
                created_at: resource.created_at,
                updated_at: resource.updated_at,
                credentials: resource.credentials
              }
            }
          end

          def set_avatar(id)
            user = User.find(id)
            avatar = rails_blob_url(user.avatar) if user.avatar.attached?

            avatar
          end
      end
    end
  end
end
