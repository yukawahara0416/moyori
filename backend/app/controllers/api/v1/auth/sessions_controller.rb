module Api
  module V1
    module Auth
      class SessionsController < DeviseTokenAuth::SessionsController
      end
    end
  end
end
