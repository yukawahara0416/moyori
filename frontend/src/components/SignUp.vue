<template>
  <div>
    <h2>SignUp</h2>
    <input type="text" v-model="signUpFormData.name" placeholder="name" />
    <input type="text" v-model="signUpFormData.email" placeholder="email" />
    <input
      type="text"
      v-model="signUpFormData.password"
      placeholder="password"
    />
    <input
      type="text"
      v-model="signUpFormData.password_confirmation"
      placeholder="password_confirmation"
    />
    <v-btn type="submit" @click="signUp">
      アカウント登録
    </v-btn>
    <div>{{ signUpFormData }}</div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
  computed: {
    ...mapGetters(['signUpFormData', 'currentUser'])
  },

  methods: {
    signUp() {
      this.$store.dispatch('signUp', this.signUpFormData)
    }
  }
}
</script>
