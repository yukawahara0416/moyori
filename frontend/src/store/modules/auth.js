import { axiosBase } from '@/plugins/axios.js'

export default {
  state: {
    currentUser: null,
    headers: null,
    signUpFormData: {
      name: '',
      email: '',
      password: '',
      password_confirmation: ''
    },
    signInFormData: {
      email: '',
      password: ''
    }
  },

  getters: {
    currentUser(state) {
      return state.currentUser
    },

    headers(state) {
      return state.headers
    },

    signInFormData(state) {
      return state.signInFormData
    },

    signUpFormData(state) {
      return state.signUpFormData
    }
  },

  mutations: {
    setCurrentUser(state, payload) {
      state.currentUser = payload.user
    },

    setSignInFormData(state, payload) {
      state.headers = {
        'access-token': payload['access-token'],
        'client': payload['client'],
        'content-type': payload['content-type'],
        'uid': payload['uid']
      }
    },

    clearSignUpFormData(state) {
      state.signUpFormData = {
        name: '',
        email: '',
        password: '',
        password_confirmation: ''
      }
    },

    clearSignInFormData(state) {
      state.signInFormData = {
        email: '',
        password: ''
      }
    },

    signIn(state, payload) {
      state.headers = {
        'access-token': payload['access-token'],
        'client': payload['client'],
        'content-type': payload['content-type'],
        'uid': payload['uid']
      }
    },

    signOut(state) {
      state.headers = null
      state.currentUser = null
    }
  },
  actions: {
    signUp(context, signUpFormData) {
      axiosBase
        .post('/api/v1/auth/', signUpFormData)
        .then(function(response) {
          context.commit('setCurrentUser', { user: response.data })
          context.commit('signIn', response.headers)
          context.commit('dialogOff', 'dialogSign')
          context.commit('clearSignUpFormData')
          context.commit('clearSignInFormData')
          context.dispatch('pushSnackbar', {
            message: 'MoYoRiへようこそ！',
            color: 'success'
          })
        })
        .catch(function() {
          context.dispatch('pushSnackbar', {
            message: 'アカウント作成に失敗しました',
            color: 'error'
          })
        })
    },

    signIn(context, signInFormData) {
      if (context.state.currentUser === null) {
        axiosBase
          .post('/api/v1/auth/sign_in', signInFormData)
          .then(function(response) {
            context.commit('setCurrentUser', { user: response.data })
            context.commit('signIn', response.headers)
            context.commit('dialogOff', 'dialogSign')
            context.commit('clearSignInFormData')
            context.commit('clearSignUpFormData')
            context.dispatch('pushSnackbar', {
              message: 'ログインしました',
              color: 'success'
            })
          })
          .catch(function() {
            context.dispatch('pushSnackbar', {
              message: 'ログインに失敗しました',
              color: 'error'
            })
          })
      } else {
        context.dispatch('pushSnackbar', {
          message: 'すでにログイン中です',
          color: 'error'
        })
      }
    },

    signOut(context) {
      axiosBase
        .delete('api/v1/auth/sign_out', {
          headers: context.state.headers
        })
        .then(function() {
          context.commit('signOut')
          context.dispatch('pushSnackbar', {
            message: 'ログアウトしました',
            color: 'success'
          })
        })
        .catch(function() {
          this.$store.dispatch('pushSnackbar', {
            message: '予期しないエラーが発生しました',
            color: 'error'
          })
        })
    },

    deleteAccount(context) {
      axiosBase
        .delete('/api/v1/auth/', { headers: context.state.headers })
        .then(function() {
          context.commit('signOut')
          context.dispatch('pushSnackbar', {
            message: 'アカウントを削除しました',
            color: 'success'
          })
          // router.push('/')
        })
        .catch(function() {
          context.dispatch('pushSnackbar', {
            message: '予期しないエラーが発生しました',
            color: 'error'
          })
        })
    }
  }
}
