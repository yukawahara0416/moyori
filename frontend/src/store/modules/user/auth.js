import { axiosBase } from '@/plugins/axios.js'
import router from '@/router'

export default {
  state: {
    currentUser: { data: {}, avatar: '' },

    headers: null,

    signUpFormData: {
      name: '',
      email: '',
      password: ''
    },

    signInFormData: {
      email: '',
      password: ''
    }
  },

  getters: {
    currentUser(state) {
      return state.currentUser
    },

    headers(state) {
      return state.headers
    },

    isLoggingIn(state) {
      return state.headers == null ? false : true
    },

    signInFormData(state) {
      return state.signInFormData
    },

    signUpFormData(state) {
      return state.signUpFormData
    }
  },

  mutations: {
    setCurrentUser(state, payload) {
      state.currentUser.data = payload
    },

    editCurrentUser(state, { name, email }) {
      state.currentUser.data.name = name
      state.currentUser.data.email = email
    },

    editCurrentUserAvatar(state, payload) {
      state.currentUser.data.avatar = payload
    },

    clearSignUpFormData(state) {
      state.signUpFormData = {
        name: '',
        email: '',
        password: ''
      }
    },

    clearSignInFormData(state) {
      state.signInFormData = {
        email: '',
        password: ''
      }
    },

    setHeaders(state, payload) {
      state.headers = {
        'access-token': payload['access-token'],
        'client': payload['client'], // eslint-disable-line
        'content-type': payload['content-type'],
        'uid': payload['uid'] // eslint-disable-line
      }
    },

    clearHeaders(state) {
      state.headers = null
      state.currentUser = { data: {}, avatar: '' }
    }
  },

  actions: {
    signUp(context, signUpFormData) {
      return axiosBase
        .post('/api/v1/auth/', signUpFormData)
        .then(response => {
          return response
        })
        .catch(() => {
          throw new Error('アカウント作成に失敗しました')
        })
    },

    signIn(context, signInFormData) {
      return axiosBase
        .post('/api/v1/auth/sign_in', signInFormData)
        .then(response => {
          return response
        })
        .catch(() => {
          throw new Error('ログインに失敗しました')
        })
    },

    signOut(context, headers) {
      return axiosBase
        .delete('api/v1/auth/sign_out', { headers })
        .then(response => {
          return response
        })
        .catch(() => {
          throw new Error('ログアウトに失敗しました')
        })
    },

    updateAccount(context, { params, headers }) {
      return axiosBase
        .patch('/api/v1/auth/', params, { headers })
        .then(response => {
          return response
          // const user = response.data.data

          // context.dispatch('editAvatar', user.id)
          // context.commit('user/editUserStore', {
          //   name: user.name,
          //   email: user.email
          // })
        })
        .catch(() => {
          throw new Error('アカウントの編集に失敗しました')
        })
    },

    deleteAccount(context, headers) {
      axiosBase
        .delete('/api/v1/auth', { headers })
        .then(() => {
          context.commit('clearHeaders')
          router.push('/')
        })
        .catch(() => {
          throw new Error('アカウントの削除に失敗しました')
        })
    },

    editAvatar(context, userId) {
      axiosBase.get('/api/v1/users/' + userId).then(response => {
        const avatar = response.data.avatar

        context.commit('editCurrentUserAvatar', avatar)
        context.commit('user/editUserAvatarStore', avatar)
      })
    },

    clearSignFormData(context) {
      context.commit('clearSignInFormData')
      context.commit('clearSignUpFormData')
    }
  }
}
