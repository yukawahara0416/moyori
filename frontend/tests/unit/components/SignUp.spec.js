import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import SignUp from '@/components/SignUp.vue'
import auth from '@/store/modules/auth.js'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuetify)

let wrapper
let store
let state
let actions
let vuetify

beforeEach(() => {
  state = {
    currentUser: {},
    signUpFormData: {
      name: 'hoge',
      email: 'hoge@example.com',
      password: 'password',
      password_confirmation: 'password'
    }
  }

  actions = {
    signUp: jest.fn()
  }

  store = new Vuex.Store({
    modules: {
      auth: {
        state,
        getters: auth.getters,
        actions
      }
    }
  })

  vuetify = new Vuetify()

  wrapper = mount(SignUp, {
    localVue,
    store,
    vuetify
  })
})

afterEach(() => {
  wrapper.destroy()
})

describe('v-on', () => {
  it('signUp', () => {
    const event = jest.fn()
    wrapper.setMethods({ signUp: event })
    wrapper.findAll('.v-btn').trigger('click')
    expect(event).toHaveBeenCalledTimes(1)
  })
})

describe('getters', () => {
  it('signUpFormData', () => {
    expect(wrapper.vm.signUpFormData).toEqual(state.signUpFormData)
  })

  it('currentUser', () => {})
})

describe('actions', () => {
  it('signUp', () => {
    wrapper.vm.signUp()
    expect(actions.signUp).toHaveBeenCalled()
  })
})

describe('template', () => {
  it('snapshot', () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
