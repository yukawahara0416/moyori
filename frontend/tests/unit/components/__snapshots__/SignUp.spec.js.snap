// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template snapshot 1`] = `
<div>
  <h2>
    SignUp
  </h2>
   
  <input
    placeholder="name"
    type="text"
  />
   
  <input
    placeholder="email"
    type="text"
  />
   
  <input
    placeholder="password"
    type="text"
  />
   
  <input
    placeholder="password_confirmation"
    type="text"
  />
   
  <button
    class="v-btn v-btn--contained theme--light v-size--default"
    type="submit"
  >
    <span
      class="v-btn__content"
    >
      
    アカウント登録
  
    </span>
  </button>
   
  <div>
    {
  "name": "hoge",
  "email": "hoge@example.com",
  "password": "password",
  "password_confirmation": "password"
}
  </div>
</div>
`;
